executable-spec/test/Cardano/Ledger/Spec/STS/Chain/Chain/Properties.hs

```diff
@@ -46,3 +53,243 @@ qc_revealsAreClassified
+         $ qc_onlyValidSignalsAreGenerated
+  where
+    maxTraceLength = 100
+
+-- Calculates a minimum required length for a trace
+-- in order to allow for stabilization of events,
+-- with input the k security parameter (number of Blocks)
+-- submit sip - 2k blocks - reveal sip - 2k blocks - active sip -
+-- voting period D blocks - 2k blocks - tally
+getMinTraceLength :: BlockCount -> Word64
+getMinTraceLength k =
+  let kval = unBlockCount k
+      minVotingDuration = unSlotCount $ Data.vpDurationToSlotCnt Data.VPMin
+  in 2*kval + 2*kval + minVotingDuration + 2*kval
+
+-- | Returns the percent of Txs with a non-empty update payload in the input Trace
+updatePayloadPct
+  :: Trace.Trace (CHAIN ShortHash)
+  -> Float
+updatePayloadPct tr =
+  let -- get the total of transactions in tr
+      txTot = foldl' (\tot b -> tot + txsInAblock b) 0 blocks
+      blocks = (Trace.traceSignals Trace.NewestFirst tr)
+      txsInAblock = \b -> length $ Body.transactions (Chain.body b)
+      -- get the total of Txs with a non-empty update payload
+      txupdTot = foldl' (\tot b -> tot + updtxsInAblock b) 0 blocks
+      updtxsInAblock = \b -> length
+                               -- Txs with a non-empty update payload
+                               $ filter (\txb -> not . null
+                                                  $ Transaction.update txb
+                                        )
+                               $ map (Transaction.body)
+                               $ Body.transactions (Chain.body b)
+
+  in (fromIntegral txupdTot) / (fromIntegral txTot) * 100
+
+submittedSIPsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+submittedSIPsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.subsips lastSt /= Map.empty
+
+revealedSIPsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+revealedSIPsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.sipdb lastSt /= Map.empty
+
+sipBallotsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+sipBallotsExist tr =
+  let lastSt = Trace.lastState tr
+  in Chain.ballots lastSt /= Map.empty
+
+voteResultsExist :: Trace.Trace (CHAIN ShortHash)  -> Bool
+voteResultsExist tr =
+  let lastSt = Trace.lastState tr
+      vResults = Map.elems $ Chain.vresips lastSt
+  in any (\vr -> Data.stakeInFavor vr /=  0
+                 && Data.stakeAgainst vr /= 0
+                 && Data.stakeAbstain vr /= 0
+         ) vResults
+
+-- Returns true if the last state of the inpuτ trace
+-- shows that all phased in the liefecycle of a software update
+-- have been covered
+lifecycleCoverage
+  :: Trace.Trace (CHAIN ShortHash)
+  -> Bool
+lifecycleCoverage tr =
+  let lastSt = Trace.lastState tr
+      subsips = Chain.subsips lastSt /= Map.empty
+      revsips = Chain.sipdb lastSt /= Map.empty
+      asips = Chain.asips lastSt /= Map.empty
+      ballots = Chain.ballots lastSt /= Map.empty
+      vresips = Chain.vresips lastSt /= Map.empty
+      apprvsips = Chain.apprvsips lastSt /= Set.empty
```

Comment:
Is it ok to check this, or do we need a check like the one in `voteResultsExist`? What's the difference between checking this and checking that there were some votes?

Reply:
we need a check like the one in `voteResultsExist`. The `vresips = Chain.vresips lastSt /= Map.empty` is a wrong check.

Reply:
Fixed.OK